### ISSUE TITLE
Complete magic methods

### ISSUE THREAD

**Issue by BladeMF** [phpactor/phpactor#1177]
In that case:
```php
function <>
```

We should complete all the 'magic' php methods and insert their signature on `tab`:
> The function names __construct(), __destruct(), __call(), __callStatic(), __get(), __set(), __isset(), __unset(), __sleep(), __wakeup(), __serialize(), __unserialize(), __toString(), __invoke(), __set_state(), __clone() and __debugInfo() are magical in PHP classes.

(from https://www.php.net/manual/en/language.oop5.magic.php)

**Comment by dantleech** [comment]
hmm, this should already work? magic methods still need to be defined?

**Comment by BladeMF** [comment]
I mean completion when defining them (as opposed to using them).

**Comment by dantleech** [comment]
Oh right -- do we want to encourage that how often would somebody do that? :smile: 

But yes maybe -- there is currently no support for completing method names, but it would be good to also have:

- Parent methods (overriding)
- Constructor
- ...

**Comment by BladeMF** [comment]
> Oh right -- do we want to encourage that - how often would somebody do that? 😄

Well, I agree with your point, but I really miss the `__construct` completion :-)

**Comment by camilledejoye** [comment]
I agree with @dantleech for the completion of methods from parents/interfaces, it could even be integrated with LSP to dispatch a command allowing to generate them automatically.
Like creating a constructor with the arguments from the parent class, overriding a method and generating the `parent::<method>(...)` instruction, etc.

The tricky part might be to deal with the detection of different cases:
- `public function __con<>`
- 'function __con<>`
- `__con`

Allowing only completion in the first case would make things easier.
In the second case we could be defining an anonymous function and completion should not be triggerd, IMO.
In the third case it might be even worse, it could be any text typed anywhere... It could be great once in a while when we want to reference a method in comment or something but most of the time I prefer my editor to do as less as possible.

**Comment by BladeMF** [comment]
> In the second case we could be defining an anonymous function and completion should not be triggerd, IMO.
We can check if that is the case. We would only complete in `ClassMembers` scope. Otherwise that's a perfectly valid case for completion IMO. That's valid for all the cases, in fact. We can be sure you are in the name of a class member, rather than a variable in a method or something else.

**Comment by camilledejoye** [comment]
It will depends on how the parse behave, it's supposed to handle invalid AST but if it can't recognize the current node as a method declaration without the rest of the code...

**Comment by BladeMF** [comment]
It can. I played with this with the `KeywordCompletor`. At least when I tested it I had pretty good knowledge of where the user was.

**Comment by dantleech** [comment]
We now provie weak support for suggestion `__construct`:
 https://github.com/phpactor/phpactor/blob/85a4be85979e1d5b930894aa90afb7721d552167/lib/Completion/Bridge/TolerantParser/WorseReflection/KeywordCompletor.php#L28-L33

But this doesn't work so well with other methods, I think I'd want to solve this by also providing methods from parent classes.
